### PR TITLE
Added backport.yml to automate backport process

### LINKED ISSUE #1112: Automate backports with a bot
Right now the backporting process requires a bunch of manual labour.

In practice, we put the backporting targets in the pull request description, so we should be able to automate this process.

### PR DESCRIPTION
**Pull request by siddharthasonker95**
Objective: Automate the backport process.

Description: 

Add the label like `backport <branch_name>` to the pull request that needs to be backported to a specific branch. Once the pull request is successfully merged, the backport workflow will create a pull request with the required changes to be merged into the selected branch.

Changes:
- Added `backport.yml` and `/scripts/backport/main.sh` to automate `backport` process. 

Resolves #1112 

## Target Release

1.8.0

### CODE REVIEW

**Comment by Yantrio** [comment]
Please dont forget to sign off on the DCO @siddharthasonker95 :)

**Comment by Yantrio** [comment]
I think it may be best to also implement the backporting from a comment, as per these docs: https://github.com/korthout/backport-action?tab=readme-ov-file#trigger-using-a-comment


also we could configure it to copy assignees/reviewers too. That would be helpful.

File: .github/workflows/backport.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      ) || (
+        github.event_name == 'issue_comment' &&
+        github.event.issue.pull_request &&
+        github.event.comment.user.id != 97796249 &&
+        startsWith(github.event.comment.body, '/backport')
+      )
+    steps:
+      - uses: actions/checkout@v4
+      - name: Create backport pull requests
+        uses: korthout/backport-action@v2
```
Comment (janosdebugs) [inline .github/workflows/backport.yml:28]: Adding GitHub Actions like this is a security risk as they get GitHub access tokens. This action isn't well known and seems to be maintained by a single person. I would advise against using it in this fashion. In fact, I would strongly prefer not to introduce any more GitHub Actions as we should actually lock down the org to prevent unintended actions from being added.

**Review by janosdebugs — CHANGES_REQUESTED** [review]

File: .github/scripts/backport/main.sh [hunk 2]
```diff
@@ -0,0 +1,90 @@
+#!/bin/bash
+
+GITHUB_TOKEN=$GITHUB_TOKEN
+
+git config --global user.email "noreply@github.com"
+git config --global user.name "GitHub Actions"
+
+git config --global credential.helper "store --file=~/.git-credentials"
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:8]: I don't believe you need to configure git credentials when on a github runner with the correct setup.

File: .github/scripts/backport/main.sh [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+git config --global credential.https://github.com.username $GITHUB_TOKEN
+
+echo "Successfully authenticated!"
+
+git checkout $GITHUB_EVENT_BRANCH
+echo "Checked out main"
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:14]: This does not match the previous command

File: .github/scripts/backport/main.sh [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+# Get backport label from all the labels of the PR
+labels=$(jq --raw-output '.pull_request.labels[].name' "$GITHUB_EVENT_PATH")
+backport_label=""
+branch_name=""
+for label in "${labels[@]}"; do
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:20]: Can github labels have spaces and does this handle that case?

File: .github/scripts/backport/main.sh [hunk 5]
```diff
@@ -0,0 +1,90 @@
+#!/bin/bash
+
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:2]: Let's include set -e so that if a command exits with a non-zero status code, the script fails.

File: .github/scripts/backport/main.sh [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    if [[ $label == "backport"* ]]; then
+        backport_label=$label
+        branch_name=${label#backport }
+        break
+    fi
+done
+echo "Label Name: $backport_label"
+
+# Checkout the version branch
+git fetch --all
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:31]: Why --all and not just the branch name?

File: .github/scripts/backport/main.sh [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+git checkout $branch_name
+echo "Version branch: $branch_name"
+
+# Checkout new backport branch
+ISSUE_NUMBER=$(jq --raw-output .pull_request.number "$GITHUB_EVENT_PATH")
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:36]: Let's do all of the issue parsing in the same location in the script for clarity.  Also, I'd prefer to do as much of this as possible in the github workflow file and not in a bash script.

Comment (cam72cam) [reply .github/scripts/backport/main.sh:36]: Also why is this uppercase when all the rest of the vars are lower case?

File: .github/scripts/backport/main.sh [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+newBranch="backport/$ISSUE_NUMBER"
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:37]: Why is this camelCase when other vars are underscore_case?

File: .github/scripts/backport/main.sh [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+git checkout -b $newBranch
+git push origin $newBranch
+echo "New backport branch: $newBranch"
+
+# Get pull request and repository info.
+pullRequestNumber=$(jq --raw-output .pull_request.number "$GITHUB_EVENT_PATH")
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:43]: Same comment as above, we should do most of this in the workflow file.

File: .github/scripts/backport/main.sh [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+     jq -r '.[].sha')
+
+# Store the commit SHAs in an array
+commit_ids_array=($commit_ids)
+for sha in "${commit_ids_array[@]}"; do
+    echo "Commit SHA: $sha"
+done
+
+# Count the number of commits
+commitCount=$(echo "$commitIDs" | wc -l)
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:63]: This is not nessesary

File: .github/scripts/backport/main.sh [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+# Checkout the version branch
+git fetch --all
+git checkout $branch_name
+echo "Version branch: $branch_name"
+
+# Checkout new backport branch
+ISSUE_NUMBER=$(jq --raw-output .pull_request.number "$GITHUB_EVENT_PATH")
+newBranch="backport/$ISSUE_NUMBER"
+git checkout -b $newBranch
+git push origin $newBranch
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:39]: Why do we push this branch before making any changes?

File: .github/scripts/backport/main.sh [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+done
+echo "Cherry Pick done!"
+
+git commit -am "Backport PR #$ISSUE_NUMBER"
+echo "Successfully committed!"
+
+git push origin $newBranch
+echo "Successfully pushed!"
+
+head_branch=$(jq -r '.pull_request.head.ref' "$GITHUB_EVENT_PATH")
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:80]: Same comment on doing this in the workflow

File: .github/workflows/backport.yml [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+      )
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v3
+    - name: Run custom bash script for backporting
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        GITHUB_EVENT: ${{ toJson(github.event) }} # Pass the github.event object as a JSON string
+      run: |
+        cd $GITHUB_WORKSPACE  # Move to the root of the repository
```
Comment (cam72cam) [inline .github/workflows/backport.yml:30]: This is not necessary

File: .github/workflows/backport.yml [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+        cd .github/scripts/backport
```
Comment (cam72cam) [inline .github/workflows/backport.yml:31]: Why cd into the scripts directory when you can `./.github/scripts/backport/main.sh` instead?

**Review by cam72cam — DISMISSED** [review]
Looks like a reasonable start, but I've added a number of comments.

File: .github/scripts/backport/main.sh (hunk 7 above)
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh:36]: I tested the commands on my terminal on personal repo and didn't notice the variable naming consistency.

File: .github/workflows/backport.yml (hunk 14 above)
Comment (siddharthasonker95) [reply .github/workflows/backport.yml:31]: I tried using `./.github/scripts/backport/main.sh` first in my repo but somehow it was not able to locate the directory.

File: .github/scripts/backport/main.sh [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+set -e
+
+# Setting up Committer identity.
+git config --global user.email "noreply@github.com"
+git config --global user.name "GitHub Actions"
+
+# Get backport label from all the labels of the PR
+backport_label=""
+branch_name=""
+for label in "${LABELS[@]}"; do
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:12]: Question for @cube2222, do we want to support backporting to multiple versions?

File: .github/scripts/backport/main.sh [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+
+# Get the commit SHAs associated with the pull request
+commit_ids=$(curl -sS -H "Authorization: token $GITHUB_TOKEN" \
+     -H "Accept: application/vnd.github.v3+json" \
+     "https://api.github.com/repos/$OWNER/$REPO/pulls/$PR_NUMBER/commits" | \
+     jq -r '.[].sha')
+
+# Store the commit SHAs in an array
+commit_ids_array=($commit_ids)
+for commit_sha in "${commit_ids_array[@]}"; do
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:41]: Why is this using an array instead of simple iteration?

File: .github/scripts/backport/main.sh [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+    git fetch origin $commit_sha
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:42]: Is this git fetch needed?

File: .github/scripts/backport/main.sh [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    echo "Cherry-picking commit: $commit_sha"
+done
+echo "Cherry Pick done!"
+
+git push origin $new_branch
+echo "Successfully pushed!"
+
+# Create the pull request.
+title="Backport PR #$ISSUE_NUMBER"
+body="This pull request backports changes from $head_branch to $branch_name."
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:53]: head_branch is not defined.

**Comment by cam72cam** [comment]
Another question for @cube2222: Do we want to cherry-pick individual commits or the merge commit?

File: .github/scripts/backport/main.sh (hunk 17 above)
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh:42]: Yes, without `git fetch`, it was not able to detect the commit ID. That's why I added `git fetch` for every commit ID for safety.

**Comment by cube2222** [comment]
@cam72cam Since we squash, I don't believe there's a distinction here, no? That is, there's just a single commit for each change.

**Comment by cam72cam** [comment]
I'm curious if it would make a merge conflict easier to spot with the merge commit vs all the branch commits.

**Comment by siddharthasonker95** [comment]
IMO, detecting merge conflicts from a merge commit will be more challenging compared to handling conflicts from individual commits one by one. Another workaround, if we choose to cherry-pick each commit for backporting, is to have the script fail when a conflict occurs during cherry-picking. Additionally, it will add a comment to the existing pull request with the commit ID that caused the conflict.
cc @cam72cam @cube2222

File: .github/scripts/backport/main.sh (hunk 18 above)
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: done.

**Comment by siddharthasonker95** [comment]
I have addressed the PR comments and added the changes where, if the backport script fails to cherry-pick a commit due to a merge conflict, it will add a comment to the existing PR, specifying the commit ID that caused the conflict.
PTAL @cam72cam

File: .github/scripts/backport/main.sh (hunk 16 above)
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: done

File: .github/scripts/backport/main.sh [hunk 19]
```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+
+set -e
+
+# Setting up Committer identity.
+git config --global user.email "noreply@github.com"
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:6]: Let's remove the --global here just in case someone runs this locally.

File: .github/workflows/backport.yml (hunk 13 above)
Comment (cam72cam) [reply .github/workflows/backport.yml:30]: @siddharthasonker95 why did you resolve this?

File: .github/scripts/backport/main.sh [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+
+# Exit if backport label not found
+echo "Label Name: $backport_label"
+if [ -z "$backport_label" ]; then
+    echo "Backport label not found. Exiting."
+    exit 1
+fi
+
+# Checkout the version branch
+git fetch origin "$branch_name"
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:29]: Why are these steps necessary?

File: .github/scripts/backport/main.sh [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+git checkout "$branch_name"
+echo "Version branch: $branch_name"
+
+# Checkout new backport branch
+new_branch="backport/$ISSUE_NUMBER"
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:34]: We should include the label in the branch name.

File: .github/scripts/backport/main.sh [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+done
+echo "Cherry Pick done!"
+
+git push origin "$new_branch"
+echo "Successfully pushed!"
+
+# Create the pull request.
+title="Backport PR #$ISSUE_NUMBER"
+body="This pull request backports changes from $HEAD_BRANCH to $branch_name."
+url=$(curl -sS -X POST -H "Authorization: token $GITHUB_TOKEN" \
```
Comment (cam72cam) [inline .github/scripts/backport/main.sh:66]: Why did you remove the usage of the `gh` command?  If you set GH_TOKEN correctly, it should work just fine.  I'd prefer that over constructing this inline.

File: .github/workflows/backport.yml (hunk 13 above)
Comment (siddharthasonker95) [reply .github/workflows/backport.yml:30]: I assumed it was related to this [comment](https://github.com/opentofu/opentofu/pull/1452/files#r1576121524) and have replied to it.

File: .github/scripts/backport/main.sh (hunk 22 above)
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh:66]: I have replaced the `gh` command changes to maintain consistency in the code, as all the other git actions have.

File: .github/scripts/backport/main.sh (hunk 21 above)
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh:34]: Sure, as the label will have format like `backport <branch_name>`, then we can name the backport branch like `backport/<branch_name>/<issue_number>`

File: .github/scripts/backport/main.sh [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+        branch_name=${label#backport }
+        break
+    fi
+done
+
+# Exit if backport label not found
+echo "Label Name: $backport_label"
+if [ -z "$backport_label" ]; then
+    echo "Backport label not found. Exiting."
+    exit 1
```
Comment (siddharthasonker95) [inline .github/scripts/backport/main.sh:25]: Should it return `exit 1` or `exit 0` when backport label is not present?
As this script runs whenever a PR is closed, and the backport label is present, then the rest of the script will run; otherwise, it will fail.
cc @janosdebugs @cam72cam

Comment (cam72cam) [reply .github/scripts/backport/main.sh:25]: exit 0

File: .github/workflows/backport.yml [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+    - name: Checkout
+      uses: actions/checkout@v3
+    - name: Set variables
+      run: |
+        echo "PR_NUMBER=${{ github.event.pull_request.number }}" >> $GITHUB_ENV
+        echo "OWNER=${{ github.event.repository.owner.login }}" >> $GITHUB_ENV
+        echo "REPO=${{ github.event.repository.name }}" >> $GITHUB_ENV
+        echo "ISSUE_NUMBER=${{ github.event.pull_request.number }}" >> $GITHUB_ENV
+        echo "HEAD_BRANCH=${{ github.event.pull_request.head.ref }}" >> $GITHUB_ENV
+        echo "LABELS=$(jq -r '.pull_request.labels[].name' "$GITHUB_EVENT_PATH")" >> $GITHUB_ENV
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/workflows/backport.yml]: I have separated setting the variable from the script execution step for better readability.

File: .github/workflows/backport.yml [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+  pull_request:
+    types:
+      - closed
+
+jobs:
+  backport:
+    name: Backport pull request
+    runs-on: ubuntu-latest
+    if: >
+      github.event.pull_request.merged && ${{contains(github.event.pull_request.labels.*.name, 'backport')}}
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/workflows/backport.yml:13]: added readme file [here.](https://github.com/opentofu/opentofu/pull/1452/commits/23d19654d9b27e11be9850409fc14c0dd5dd99b4)

File: .github/workflows/backport.yml (hunk 14 above)
Comment (siddharthasonker95) [reply .github/workflows/backport.yml]: updated.

File: .github/scripts/backport/main.sh [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+
+# Get the commit SHAs associated with the pull request
+gh api "/repos/$OWNER/$REPO/pulls/$PR_NUMBER/commits" --jq '.[].sha' |
+while IFS= read -r commit_sha; do
+    git fetch origin "$commit_sha"
+    if git cherry-pick "$commit_sha" ; then
+        echo "Cherry-picking commit: ${commit_sha}"
+    else
+        # Add a failure comment to the pull request
+        echo "Error: Failed to cherry-pick commit ${commit_sha}"
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/8f625a42772194df421f5d5eb45eefb9ba8fc201)

File: .github/scripts/backport/main.sh [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+fi
+
+# Check if gh is logged in
+if ! gh auth status >/dev/null 2>&1; then
+    echo "Error: Authentication check for 'gh' failed."
+    exit 1
+fi
+
+# Checkout the version branch
+echo "Fetching branch ${branch_name}..."
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: I have added the info and error logs in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/8f625a42772194df421f5d5eb45eefb9ba8fc201)

File: .github/scripts/backport/main.sh [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+    if [[ $label == "backport"* ]]; then
+        backport_label=$label
+        branch_name=${label#backport }
+        break
+    fi
+done
+
+# Exit if backport label not found
+if [ -z "$backport_label" ]; then
+    echo "Warning: The backport label is not found on the Pull Request. Skipping the backport process"
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: done.

File: .github/scripts/backport/main.sh [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+done
+echo "Cherry-picking process completed successfully for pull request #${PR_NUMBER}."
+
+git push origin "$new_branch"
+echo "Successfully pushed changes to the ${new_branch}"
+
+# Create the pull request.
+title="Backport PR #$ISSUE_NUMBER"
+body="This pull request backports changes from $HEAD_BRANCH to $branch_name."
+url=$(gh pr create --title "$title" --body "$body" --base "$branch_name" --head "$new_branch")
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: handled the error in the latest [commit.](https://github.com/opentofu/opentofu/pull/1452/commits/8f625a42772194df421f5d5eb45eefb9ba8fc201)

File: .github/scripts/backport/main.sh [hunk 30]
```diff
@@ -0,0 +1,79 @@
...
+        # Add a failure comment to the pull request
+        echo "Error: Failed to cherry-pick commit ${commit_sha}"
+        failureComment="Error: Cherry-picking commit $commit_sha failed during the backport process to $branch_name. Please resolve any conflicts and manually apply the changes."
+        gh pr comment "$PR_NUMBER" --body "$failureComment"
+        exit 1
+    fi
+done
+echo "Cherry-picking process completed successfully for pull request #${PR_NUMBER}."
+
+git push origin "$new_branch"
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: I have handled the error in every git commands in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/8f625a42772194df421f5d5eb45eefb9ba8fc201)

File: .github/scripts/backport/main.sh [hunk 31]
```diff
@@ -0,0 +1,147 @@
+#!/bin/bash
+
+set -e
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

File: .github/scripts/backport/main.sh [hunk 32]
```diff
@@ -0,0 +1,147 @@
...
+    if [ -z "$1" ]; then
+        echo "No message provided."
+        return
+    fi
+
+    echo "$1"
+}
+
+# Setting up Committer identity.
+log_info "Setting up Committer identity..."
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

File: .github/scripts/backport/main.sh [hunk 33]
```diff
@@ -0,0 +1,147 @@
...
+set -e
+
+# Function to handle errors
+log_error() {
+    if [ -z "$1" ]; then
+        echo "Error: No error message provided." >&2
+        exit 1
+    fi
+
+    echo "Error: $1" >&2
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

File: .github/scripts/backport/main.sh [hunk 34]
```diff
@@ -0,0 +1,147 @@
+#!/bin/bash
+
+set -e
+
+# Function to handle errors
+log_error() {
+    if [ -z "$1" ]; then
+        echo "Error: No error message provided." >&2
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

File: .github/scripts/backport/main.sh [hunk 35]
```diff
@@ -0,0 +1,147 @@
...
+    fi
+
+    echo "Error: $1" >&2
+    exit 1
+}
+
+# Function to output info logs
+log_info() {
+    if [ -z "$1" ]; then
+        echo "No message provided."
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

File: .github/scripts/backport/main.sh [hunk 36]
```diff
@@ -0,0 +1,147 @@
...
+        return
+    fi
+
+    echo "$1"
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: fixed in the latest [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

File: .github/scripts/backport/main.sh [hunk 37]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+# Setting up Committer identity.
+log_info "Setting up Committer identity..."
+if ! git config user.email "noreply@github.com"; then
+    log_error "Failed to set email."
+fi
+
+if ! git config user.name "GitHub Actions"; then
+    log_error "Failed to set username"
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: fixed

File: .github/scripts/backport/main.sh [hunk 38]
```diff
@@ -0,0 +1,147 @@
...
+if [ -z "$GITHUB_TOKEN" ]; then
+    log_error "GitHub token is not available. Please ensure a secret named 'GITHUB_TOKEN' is defined."
+fi
+
+# Check if gh is logged in
+if ! gh auth status >/dev/null 2>&1; then
+    log_error "Authentication check for 'gh' failed."
+fi
+
+if [ -z "$OWNER" ] || [ -z "$REPO" ] || [ -z "$PR_NUMBER" ] || [ -z "$ISSUE_NUMBER" ] || [ -z "$HEAD_BRANCH" ]; then
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: fixed in the latest commit.

File: .github/scripts/backport/main.sh [hunk 39]
```diff
@@ -0,0 +1,147 @@
...
+fi
+
+# Checkout new backport branch
+new_branch="backport/$branch_name/$ISSUE_NUMBER"
+log_info "Checking out new backport branch ${new_branch}..."
+if ! git checkout -b "$new_branch"; then 
+    log_error "Failed to create new backport branch ${new_branch}"
+fi
+
+if ! git push origin "$new_branch"; then
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: added.

File: .github/workflows/backport.yml [hunk 40]
```diff
@@ -0,0 +1,28 @@
+name: Backport Merged Pull Request
+
+on:
+  pull_request:
+    types:
+      - closed
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/workflows/backport.yml:6]: I have updated the backport.yml to trigger the backport script when a pull request is closed or labeled and pull request should fulfill following condition:
1. Pull request should be closed
2. Pull request should be merged
3. Pill request should have a label containing `backport` string.

File: .github/scripts/backport/main.sh [hunk 41]
```diff
@@ -0,0 +1,147 @@
...
+fi
+log_info "Successfully found backport label: $backport_label"
+
+# Check if GitHub token is set
+if [ -z "$GITHUB_TOKEN" ]; then
+    log_error "GitHub token is not available. Please ensure a secret named 'GITHUB_TOKEN' is defined."
+fi
+
+# Check if gh is logged in
+if ! gh auth status >/dev/null 2>&1; then
```
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/scripts/backport/main.sh]: fixed in this [commit](https://github.com/opentofu/opentofu/pull/1452/commits/b6fb1156b4f6e9858da059d5e10439e3a6257d72)

File: .github/workflows/backport.yml (hunk 40 above)
[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/workflows/backport.yml:6]: Below are the improvements that can be made to ensure the script's idempotence:

1. Validate whether the branch exists before creating it.
2. Check if the pull request already exists before attempting to create it.
3. Verify if a comment has been added to the pull request before attempting to add another.
4. Validate whether the branch already contains cherry-picked commits.

All validations are straightforward except for validating cherry-picked commits. If you have any suggestions on how to perform this validation, please let me know.

[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/workflows/backport.yml:6]: I have implemented the suggested checks. Specifically, I added a check for branch existence; if the branch exists, it deletes the branch with all temporary changes, recreates it, and restarts the backport process.

I chose not to use `git rebase` because it is more error-prone and can potentially cause conflicts or unintended changes. Instead, I opted for branch deletion and recreation to ensure a clean and stable backport process.

[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply .github/workflows/backport.yml:6]: Additionally, I included a check to verify if a pull request exists between the target branch and the new backport branch. If a pull request exists, the backport process is skipped; otherwise, it continues. 
I also added a manual test case as suggested.
PTAL again.

**Review by Andrew-Hayes — APPROVED** [review]
